gateway/src/apicast/policy/scope_check/scope_check.lua

```diff
@@ -0,0 +1,105 @@
+local policy = require('apicast.policy')
+local _M = policy.new('Scope Check Policy')
+
+local ipairs = ipairs
+local MappingRule = require('apicast.mapping_rule')
+
+local new = _M.new
+
+local function init_mapping_rules(scopes)
+  for _, scope in ipairs(scopes) do
+    scope.mapping_rule = MappingRule.from_proxy_rule({
+      http_method = 'GET',
+      pattern = scope.resource,
+      querystring_parameters = {},
+      metric_system_name = 'hits',
+      delta = 1
+    })
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.type = config.type or "whitelist"
+  self.scopes = config.scopes or {}
+
+  init_mapping_rules(self.scopes)
+
+  return self
+end
+
+local function check_roles_in_token(role, roles_in_token)
+  local matched_each_role = false
+  for _, role_in_token in ipairs(roles_in_token) do
+    if role == role_in_token then
+      matched_each_role = true
+      break
+    end
+  end
+
+  return matched_each_role
+end
+
+local function scope_check(scopes, jwt)
+  local uri = ngx.var.uri
+
+  for _, scope in ipairs(scopes) do
+    -- resource check
```

Comment:
I think it would be great to extract this part into a function. This one is too large and does two things.

In the end if it is split into more functions it could be simplified to:

```lua
for _, scope in ipairs(scopes) do
  if match_resource_roles(scope, jwt) and match_client_roles(scope, jwt) then
   return true
  end
end
```

Reply:
Changed.